## DANNY MAS SQL PROJECT WEEK 1

### QUESTION 1
#### WHAT IS THE TOTAL AMOUNT EACH CUSTOMER SPENT AT THE RESTAURANT?

In [3]:
%%sql
select customer_id, sum(price) as total_amount_spent
from sales
inner join menu
using (product_id)
group by customer_id
order by total_amount_spent desc;

 * postgresql://postgres:***@localhost/postgres
3 rows affected.


[('A', 76), ('B', 74), ('C', 36)]

>The information requested by the first question is embedded in more than one table, therefore I had to employ the use of "join" which is a function used to merge tables together in sql. Since we are looking for the total amount spent, I summed up the amount spent each day for each customer hence also the use of group by clause.
I further used the order by clause as to enhance ease of reading. 

>From the above result, customer A spent the highest amount of money 76 dollars purchasing products from the store, followed closely by customer B who spent 74 dollars and then customer C spending the least 36 dollars

### QUESTION 2
#### HOW MANY DAYS HAS EACH CUSTOMER VISITED THE RESTAURANT?

In [4]:
%%sql
-- From the dataset, we have customers visiting the store more than once a day, so the need to count each day as unique 

select customer_id, count( distinct order_date) as number_of_days_visited
from sales
group by  customer_id

 * postgresql://postgres:***@localhost/postgres
3 rows affected.


[('A', 4), ('B', 6), ('C', 2)]

>Customer B has visited the restaurant the highest number of times. Though customer A and customer C had visited lesser times, they at one point visited the restaurant more than once in a day. This is the reason i used a distinct clause in my query as to ensure I capture the accurate number of days.

### QUESTION 3

#### WHAT WAS THE FIRST ITEM FROM THE MENU PURCHASED BY EACH CUSTOMER?

In [5]:
%%sql
select customer_id, product_name 
from
(select customer_id, product_name,
dense_rank () over(partition by customer_id order by order_date) first_item
from sales
inner join menu
using (product_id)) as purchases
where first_item = 1
group by customer_id,product_name

 * postgresql://postgres:***@localhost/postgres
4 rows affected.


[('A', 'curry'), ('A', 'sushi'), ('B', 'curry'), ('C', 'ramen')]

>Given the abscence of order time in the dataset, the syntax above returns the purchases made on the first day for each customer. As noted earlier, customer A and customer C purchased more than once on their first day of purchase, thus we have two products listed for them both.

>Customer A's first purchases were Curry and Sushi.
>Customer B's first purchase was Curry while
>Customer C's first purchase was Ramen. 

>It is necessary to state here that customer C purchased Ramen twice on the first day thus the reason I grouped by the product name as to avoid repetition of entries

### QUESTION 4
#### WHAT IS THE MOST PURCHASED ITEM ON THE MENU AND HOW MANY TIMES WAS IT PURCHASED BY ALL CUSTOMERS?

In [6]:
%%sql
select product_name, count(product_name) as most_purchased from sales
inner join menu using (product_id)
group by product_name
order by count(product_name) desc
limit 1

 * postgresql://postgres:***@localhost/postgres
1 rows affected.


[('ramen', 8)]

> Of the products purchased, Ramen was purchased the most by all customers. The restaurant must be really good at making Ramen ):

### QUESTION 5 
#### WHICH ITEM WAS THE MOST POPULAR FOR EACH CUSTOMER?

In [7]:
%%sql

select customer_id, product_name, count(product_name) as number_of_products,
dense_rank () over(partition by customer_id order by count(product_name) desc) rank
from sales 
inner join menu using (product_id)
group by customer_id, product_name

 * postgresql://postgres:***@localhost/postgres
7 rows affected.


[('A', 'ramen', 3, 1),
 ('A', 'curry', 2, 2),
 ('A', 'sushi', 1, 3),
 ('B', 'sushi', 2, 1),
 ('B', 'curry', 2, 1),
 ('B', 'ramen', 2, 1),
 ('C', 'ramen', 3, 1)]

>The above result shows the number of time each product was purchased by each customer and their respective rankings, with the product purchased most having 1 as rank since from the query, I ranked them in descending order. This query would then be merged in another query as to effectively answer the poised question.

In [8]:
%%sql
select customer_id, product_name, number_of_products from (
select customer_id, product_name, count(product_name) as number_of_products,
dense_rank () over(partition by customer_id order by count(product_name) desc) popular_item
from sales 
inner join menu using (product_id)
group by customer_id, product_name) as subquery
where popular_item = 1

 * postgresql://postgres:***@localhost/postgres
5 rows affected.


[('A', 'ramen', 3),
 ('B', 'sushi', 2),
 ('B', 'curry', 2),
 ('B', 'ramen', 2),
 ('C', 'ramen', 3)]

>Limiting the result to only rows with ranks as 1 answers the question of the most popular item for each customer

>It seems Customer B values all product the same as she buys the same number for all products within the timeframe analysed.

>While for Customer A and Customer C we get to see that they both enjoy Ramen the most and thus Ramen is the most popular for them.

### QUESTION 6 
#### WHICH ITEM WAS PURCHASED FIRST BY THE CUSTOMER AFTER THEY BECAME A MEMBER?

In [9]:
%%sql
select customer_id,product_name,order_date,
dense_rank () over(partition by customer_id order by order_date) as rankings
from sales 
inner join menu using (product_id)
inner join members using (customer_id)
where order_date >= join_date


 * postgresql://postgres:***@localhost/postgres
7 rows affected.


[('A', 'curry', datetime.date(2021, 1, 7), 1),
 ('A', 'ramen', datetime.date(2021, 1, 10), 2),
 ('A', 'ramen', datetime.date(2021, 1, 11), 3),
 ('A', 'ramen', datetime.date(2021, 1, 11), 3),
 ('B', 'sushi', datetime.date(2021, 1, 11), 1),
 ('B', 'ramen', datetime.date(2021, 1, 16), 2),
 ('B', 'ramen', datetime.date(2021, 2, 1), 3)]

>This shows the ranking of all products purchased by the customers. We can se here that there is no record for Customer C, this is because from the dataset given, Customer C did not become a member 

In [10]:
%%sql
select customer_id,product_name,order_date from(
select customer_id,product_name,order_date,
dense_rank () over(partition by customer_id order by order_date) as rankings
from sales 
inner join menu using (product_id)
inner join members using (customer_id)
where order_date >= join_date) as first_purchase
where rankings = 1

 * postgresql://postgres:***@localhost/postgres
2 rows affected.


[('A', 'curry', datetime.date(2021, 1, 7)),
 ('B', 'sushi', datetime.date(2021, 1, 11))]

>Since rank 1 indicates the first time each Customer purchased a product after becoming a member, I limited the result to only records where ranking is 1

> Customer A bought curry on the same day he became a member. 
> Customer B bought sushi also on the same day he became a member.

### QUESTION 7 
#### WHICH ITEM WAS PURCHASED JUST BEFORE THE CUSTOMER BECAME A MEMBER?

In [11]:
%%sql
select distinct customer_id,product_name,order_date from(
select customer_id,product_name,order_date,
dense_rank () over(partition by customer_id order by order_date) as rankings
from sales 
inner join menu using (product_id)
inner join members using (customer_id)
where order_date < join_date) as first_purchase
where rankings = 1

 * postgresql://postgres:***@localhost/postgres
3 rows affected.


[('A', 'curry', datetime.date(2021, 1, 1)),
 ('A', 'sushi', datetime.date(2021, 1, 1)),
 ('B', 'curry', datetime.date(2021, 1, 1))]

>In question 6 we selected the product bought first after becoming a member, this question addresses the product bought just before becoming a member, so there syntax is pretty similar.
The difference being that the order_date of the dense_rank function was in question 6 ordered in descending order while it took the default form(ascending order) in this question.

> Customer A bought twice on the day before he became a member thus two records are shown for Customer A, showing the purchase of Curry and Sushi
> Customer B bought curry before joining as member

### QUESTION 8
#### WHAT IS THE TOTAL ITEMS AND AMOUNT SPENT FOR EACH MEMBER BEFORE THEY BECAME A MEMBER?

In [12]:
%%sql
select customer_id, count(product_name) as total_items, sum(price) as amt_spent
from sales
inner join menu using (product_id)
inner join members using (customer_id)
where order_date < join_date
group by customer_id

 * postgresql://postgres:***@localhost/postgres
2 rows affected.


[('B', 3, 40), ('A', 2, 25)]

> Two of the customers ended up becoming joining as members, since there is a 'before becoming a member' clause in the question we are concerned with custoemr A and B.
Customer B spent 40 dollars on 3 products while Customer A spent 25 dollars on two products

### QUESTION 9
####  IF EACH \\$1 SPENT EQUATES TO 10 POINTS AND SUSHI HAS 2X POINTS MULTIPLIER-HOW MANY POINTS WOULD EACH CUSTOMER GET?

In [13]:
%%sql
select customer_id, product_name, sum(price) as price,
case when product_name = 'sushi' then sum(price) * 20
else sum(price) * 10
end as points
from sales
inner join menu using (product_id)
group by 1,2

 * postgresql://postgres:***@localhost/postgres
7 rows affected.


[('B', 'sushi', 20, 400),
 ('A', 'curry', 30, 300),
 ('C', 'ramen', 36, 360),
 ('A', 'sushi', 10, 200),
 ('B', 'curry', 30, 300),
 ('A', 'ramen', 36, 360),
 ('B', 'ramen', 24, 240)]

> This question is a bit tricky, requires the knowlege of common maths to solve, so wherever the product name is 'sushi' we multiply by 20 points and 10 points for other product_names. The first step is to sum up for each customer the amount spent buying each product and then apply the multiplier as seen above

In [14]:
%%sql
select customer_id, sum(points) as total_points from(
select customer_id, product_name, sum(price) as price,
case when product_name = 'sushi' then sum(price) * 20
else sum(price) * 10
end as points
from sales
inner join menu using (product_id)
group by 1,2) as dollar_spent
group by 1

 * postgresql://postgres:***@localhost/postgres
3 rows affected.


[('B', Decimal('940')), ('C', Decimal('360')), ('A', Decimal('860'))]

> From the question, we were asked ho many points each customer have limiting us to two columns i.e customer and points columns. This we can only get from the temporary table created above by summing up all points for each customers.

> Customer B has the highest of 940 dollars while customer C has the lowest of 360 dollars.

### QUESTION 10

#### IN THE FIRST WEEK AFTER A CUSTOMER JOINS THE PROGRAM(INCLUDING THEIR JOIN DATE)THEY EARN 2XPOINTS ON ALL ITEMS, NOT JUST SUSHI- HOW MANY POINTS DO CUSTOMER A AND B HAVE AT THE END OF JANUARY?

In [15]:
%%sql
select customer_id,order_date,join_date,join_date + interval '6day' as first_week,price,product_name
from sales
inner join menu using (product_id)
inner join members using (customer_id)
where order_date <= '2021-01-31'

 * postgresql://postgres:***@localhost/postgres
11 rows affected.


[('A', datetime.date(2021, 1, 1), datetime.date(2021, 1, 7), datetime.datetime(2021, 1, 13, 0, 0), 15, 'curry'),
 ('A', datetime.date(2021, 1, 10), datetime.date(2021, 1, 7), datetime.datetime(2021, 1, 13, 0, 0), 12, 'ramen'),
 ('A', datetime.date(2021, 1, 11), datetime.date(2021, 1, 7), datetime.datetime(2021, 1, 13, 0, 0), 12, 'ramen'),
 ('A', datetime.date(2021, 1, 11), datetime.date(2021, 1, 7), datetime.datetime(2021, 1, 13, 0, 0), 12, 'ramen'),
 ('A', datetime.date(2021, 1, 1), datetime.date(2021, 1, 7), datetime.datetime(2021, 1, 13, 0, 0), 10, 'sushi'),
 ('A', datetime.date(2021, 1, 7), datetime.date(2021, 1, 7), datetime.datetime(2021, 1, 13, 0, 0), 15, 'curry'),
 ('B', datetime.date(2021, 1, 11), datetime.date(2021, 1, 9), datetime.datetime(2021, 1, 15, 0, 0), 10, 'sushi'),
 ('B', datetime.date(2021, 1, 4), datetime.date(2021, 1, 9), datetime.datetime(2021, 1, 15, 0, 0), 10, 'sushi'),
 ('B', datetime.date(2021, 1, 2), datetime.date(2021, 1, 9), datetime.datetime(2021, 1, 15, 0, 0), 15, 'curry'),
 ('B', datetime.date(2021, 1, 1), datetime.date(2021, 1, 9), datetime.datetime(2021, 1, 15, 0, 0), 15, 'curry'),
 ('B', datetime.date(2021, 1, 16), datetime.date(2021, 1, 9), datetime.datetime(2021, 1, 15, 0, 0), 12, 'ramen')]

>This was certainly tricky for me. I employed the use of Common Table Expression (CTE). This creates a temporal table just like we have been creating through the 'from' clause.
> I created a new column which has the days within the first week a customer joined as member and also limited the order date to just January as requested in the question.

In [16]:
%%sql
with points as (
select customer_id,order_date,join_date,join_date + interval '6day' as first_week,price,product_name
	from sales
inner join menu using (product_id)
inner join members using (customer_id)
where order_date <= '2021-01-31')
select customer_id,
sum(case when order_date between join_date and first_week then price * 20
when product_name = 'sushi' then price * 20
else price * 10
end) as total_points
from points
group by customer_id

 * postgresql://postgres:***@localhost/postgres
2 rows affected.


[('A', 1370), ('B', 820)]

> For all products bought in the first week a customer became a member, they earned 20 points, all sushi earned 20 points irrespective of when they were bought while the rest of the products on every other day(aside first week of course) earns 10 points. I used 'case syntax' to ensure all these are accounted for, since we are looking for the total_point for all customer, I summed the case result.
> Customer A earned a total of 1370 points while customer B earned 820 points.